### PR TITLE
Documentation PR for Sinco v4.x

### LINKED ISSUE #88: Docs for Sinco 4.x
## Summary

What:

Documentation for Sinco 4.x

Why:

So that new stuff in Sinco 4.x is documented

### PR DESCRIPTION
**Pull request by SnoCold**
Closes #88, #72 

## Summary

* Added documentation for new stuff
*  Modified previous docs for enhancements

### CODE REVIEW

File: docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+There are a few things you would need to change, but we hoped that changing
+these would improve the experience long term, so whilst migrating is a pain (as
+we all know), the API will make more sense and be easier to write.
+
+This version, we continue to be even closer to the browser APIs. All the classes
+introduced in version 3 are carried forward, plus, we have introduced changes in
+the actual way you would interact with your browser. For instance now the
+`buildFor` method returns an object with two values - a `Client` object and a
+`Page` object, representing your browser and its first tab.
```
Comment (ebebbington) [inline docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md:13]: really like this wording

**Review by ebebbington — CHANGES_REQUESTED** [review]
Don't suppose we could also keep some naming consistencies? for example changing `const { Chrome. chromePage } = ...` to `const { chrome, chromePage } = ...`

It's really just for a clearer UX (we really strive to have the best possible UX for the website)

Also noticed the `Page#close` method isn't documented, don't suppose you could add that too? ii don't think it'll get used often (if ever), but it's a part of the API now, this method essentially just closes the page, nto allowing it to be used again

File: docs/sinco/v4.x/2_tutorials/2_client/3_work_with_tabs.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+1. Create your `app_test.ts` file.
+
+   ```typescript
+   // app_test.ts
+
+   import { assertEquals, buildFor } from "./deps.ts";
+
+   Deno.test("My web app works as expected", async () => {
+     const { browser, page } = await buildFor("chrome");
+     const { browser, page } = await buildFor(browserItem.name);
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/2_tutorials/2_client/3_work_with_tabs.md:51]: Ohhh man, 😭 , Adding this to embarrasing_stuffs_in_my_life.ts

**Review by ebebbington — APPROVED** [review]

File: docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+## `--allow-net`
+
+**Browser**: Both
+
+The `--allow-net` permission is required for Sinco to interact with the Chrome
+Devtools Protocol (CDP) via a WebSocket that is opened when a headless browser
+instance is created.
+
+As such, restricting this permission is purely a matter of specifying a hostname
+and port, both of which make up the endpoint of the websocket URL.
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md:14]:
```suggestion
and port, both of which make up the endpoint of the WebSocket URL.
```

**Review by crookse — CHANGES_REQUESTED** [review]
most of it looks good! pretty sure a lot of my comments were things i missed when reviewing v3 docs

File: docs/sinco/v4.x/1_getting_started/2_about_sinco.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+## Suggestions
+
+1. Using Chrome. While we have no preference on what you wish to use, and we
+   won't force you to use a certain browser, it's worth noting that running a
+   Firefox headless instance and communicating it via the Chrome Devtools
+   Protocol (CDP) can be around 50-66% slower than using Chrome. This would mean
+   that generally speaking, each test case (if you are using Sinco to run tests)
+   would take 50% longer if the browser chosen is Firefox. This is why we would
+   suggest Chrome if you have no specific preferences, as it will drastically
+   speed up your tests.
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/1_getting_started/2_about_sinco.md:61]: I'm on it. Will quickly resolve this one.

File: docs/sinco/v4.x/1_getting_started/3_about_the_tutorials.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+- After you follow the Steps section, you can use this Verification section to
+  test that your application is working.
+
+- Most, if not all verification steps will mention to test a file. You should
+  note that the commands these use do not restrict permissions, but it is
+  recommended that you do. The reason they don't is because what Sinco needs
+  access to is very dynamic, and very dependant on which OS and browser you are
+  using. You should refer to
+  [restricting permissions](https://drash.land/sinco/v4.x/tutorials/restricting-permissions)
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/1_getting_started/3_about_the_tutorials.md:33]: While at it, will use this too

File: docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md:13]: Copied this and applied deno fmt

File: docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+navigations, or an option during actions, such as a click.
+
+This method was used if you clicked an element and maybe it sent you to another
+page. Now, it is replaced by `.click({ waitFor: "navigation" })`. The first
+parameter, if `options.waitFor` is set to `"navigation"`, the element will be
+clicked and then it will wait for the page to change.
+
+## `Page#assertSee()`
+
+Has now been retired.
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md:73]: Retired seems more formal I suppose 😅 ?
Removed feels harsh to those methods. 🤣 
What do you think?

File: docs/sinco/v4.x/2_tutorials/2_client/3_working_with_tabs.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+Sinco provides the method `.page()` on the `Client` class that will try and
+retrieve the page (tab), of which the index (_starting at 1 for the first tab_)
+is provided as an argument. If the provided index is not within the range of
+total open tabs in the browser, Sinco will close all resources and throw a
+`RangeError`.
+
+Sinco also provides the method `closeAllPagesExcept()` on the `Client` class,
+that will try and close all open pages except the one passed in as the argument.
+Please be aware that any other page objects will then not allow interaction with
+their respective page and any such attempt will result in an Error.
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/2_tutorials/2_client/3_working_with_tabs.md:21]: Definitely, will update this

File: docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+The hostname defaults to "localhost", and the port defaults to "9292", so
+restricting the permissions would look like:
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md:17]: Definitely. Will resolve this

File: docs/sinco/v4.x/2_tutorials/6_using_within_docker.md [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    docker-compose.yml
+    app.dockerfile
+    ▾ src/
+        app_test.ts
+        deps.ts
+```
+
+## Steps
+
+1. Create your dockerfile.
```
[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/2_tutorials/6_using_within_docker.md:45]: Definitely, resolving this as well

File: docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (crookse) [reply docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md]: haha. that's true. retired does seem more formal. whatever you choose is fine with me honestly. i think the important part is we be consistent in migration guides.

[the remark this answers is not in the document]
Comment (SnoCold) [reply docs/sinco/v4.x/1_getting_started/4_v3_to_v4_migration_guide.md]: then I shall make it removed, no issues at all, I have already pushed it. 😄

**Review by crookse — APPROVED** [review]

**Comment by crookse** [comment]
@SnoCold LGTM. im still in the process of reviewing v4, but these docs are good as of now. awesome work!

File: docs/sinco/v4.x/1_getting_started/2_about_sinco.md [hunk 11]
```diff
@@ -58,4 +59,4 @@ Firefox headless instance and communicating it via the Chrome Devtools Protocol
 generally speaking, each test case (if you are using Sinco to run tests) would
 take 50% longer if the browser chosen is Firefox. This is why we would suggest
 Chrome if you have no specific preferences, as it will drastically speed up your
-tests.
+tests. -->
```
Comment (crookse) [inline docs/sinco/v4.x/1_getting_started/2_about_sinco.md:62]: what's the --> here ?

Comment (crookse) [reply docs/sinco/v4.x/1_getting_started/2_about_sinco.md:62]: nvm... scrolled too fast

File: docs/sinco/v4.x/2_tutorials/2_client/1_creating_a_client.md [hunk 12]
```diff
@@ -65,11 +65,11 @@ In this tutorial, you will:
    import { buildFor } from "./deps.ts";
 
    // `buildFor()` takes two arguments:
-   //   1. The browser name to build for. This can be "firefox" or "chrome". This is required.
+   //   1. The browser name to build for. This can be "chrome". This is required.
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/2_client/1_creating_a_client.md:68]: should we say: currently, it only accepts "chrome" and is required?

File: docs/sinco/v4.x/2_tutorials/2_client/4_closing_the_client.md [hunk 13]
```diff
@@ -22,7 +22,7 @@ can even provide a second argument defining the class of the error. Default is
 
 In this tutorial, you will:
 
-- Create both a headless browser instance for Chrome and Firefox; and
+- Create both a headless browser instance for Chrome; and
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/2_client/4_closing_the_client.md:25]: should these bullets be:

- Create a headless browser for Chrome; and
- Close it
?

File: docs/sinco/v4.x/2_tutorials/2_client/4_closing_the_client.md [hunk 14]
```diff
@@ -43,18 +43,16 @@ In this tutorial, you will:
...
-   const { browser: Firefox, page: Fpage } = await buildFor("firefox");
    // Now you close the connections and processes, as you are done here
    await Chrome.close(); // or await Chrome.close("Some Error occured") to close and throw an Error
-   await Firefox.close(); // or await Firefox.close("The range is too big!", RangeError) to close and throw a RangeError
    ```
 
-Here you are going to create your headless browser instance for Firefox and
-Chrome. Note that you would not need to create both at the same time, you are
+Here you are going to create your headless browser instance for Chrome.
+Note that you would not need to create both at the same time, you are
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/2_client/4_closing_the_client.md:51]: doesn't seem like we need this note now

File: docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md [hunk 15]
```diff
@@ -47,8 +47,7 @@ $ deno test --allow-read="C:\Program Files\Google\Chrome\Application\chrome.exe,
 
 ## `--allow-run`
 
-This permission is required because Sinco runs a subprocess for Chrome or
-Firefox, to start a headless browser instance. So the argument to pass to this
+This permission is required because Sinco runs a subprocess for Chrome, to start a headless browser instance. So the argument to pass to this
 would be the path to your browsers executable:
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md:51]: should this be "browser's executable" ?

File: docs/sinco/v4.x/2_tutorials/6_using_within_docker.md [hunk 16]
```diff
@@ -79,7 +67,7 @@ In this tutorial, you will:
    ```
 
 Here, you are using a very small image (`debian-slim`) as your baseline for your
-Docker container. Then you install the chrome-driver and firefox-driver, which
+Docker container. Then you install the chrome-driver, which
 allows Sinco to create a headless browser instance for these browsers. Then you
```
Comment (crookse) [inline docs/sinco/v4.x/2_tutorials/6_using_within_docker.md:71]: allows Sinco to create a headless browser instance for Chrome

File: docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md (hunk 15 above)
Comment (ebebbington) [reply docs/sinco/v4.x/2_tutorials/5_restricting_permissions.md:51]: hmm i think if we getting into the grips of it, i dont think so?

like `browser's executable` is like saying "browser es executable", whilst i guess the executable word belongs to the "browser", it dont sound right? idk lol... BRENO where u at my fam

**Review by crookse — APPROVED** [review]

**Pull request merged**